### PR TITLE
Contract Owner Library

### LINKED ISSUE #53: Contract Owner Library
#### Description

In Solidity, a common library that is used is the `ownerOf` modifier. This allows for developers to ensure a function cannot be called unless a specific user is the caller.

Sway should have a similar feature that easily allows for developers to setup an owner of a contract.

#### Motivation

Contracts often require owner privileges.

### PR DESCRIPTION
**Pull request by bitzoic**
## Type of change

- New feature

## Changes

The following changes have been made:

- Added a small library to import and use ownership in a contract, the following functions have been added:
    - `only_owner()`
    - `owner()`
    - `renounce_ownership()`
    - `set_ownership()`
    - `transfer_ownership()`
- Added README.md for this library
- Added SPECIFICATION.md for this library 
- Added tests for this library

## Notes

- It may be useful to update some of the apps in the sway-applications repo to use this library, using the `as` keyword to change "owner" to "admin" as necessary

## Related Issues

Closes #53

### CODE REVIEW

File: sway_libs/src/ownership/ownable.sw
Comment (matt-user) [inline sway_libs/src/ownership/ownable.sw:45]:
```suggestion
    only_owner();
    let previous_owner = get::<Option<Identity>>(OWNER);
```
saves a storage read in the case the caller is not the owner

File: sway_libs/src/ownership/ownable.sw
Comment (matt-user) [inline sway_libs/src/ownership/ownable.sw:26]:
```suggestion
    only_owner();
    let previous_owner = get::<Option<Identity>>(OWNER);
```
saves a storage read in the case the caller is not the owner

File: sway_libs/src/ownership/ownable.sw
Comment (matt-user) [inline sway_libs/src/ownership/ownable.sw:24]: I don't see the purpose of this function.  If an owner renounces ownership of a contract then anyone can call the `set_ownership` function and gain restricted access to the contract in question.  IMO `renounce_ownership` should either set the owner to the 0 address and render all restricted access to the contract useless (if we do this we should add warnings to the code and docs).  Or `renounce_ownership` should be removed and restricted access can only be changed by using the `transfer_ownership` function

**Review by matt-user — CHANGES_REQUESTED** [review]
I have an issue with the `renounce_ownership` function but other than that it lgtm

File: sway_libs/src/ownership/ownable.sw
Comment (bitzoic) [reply sway_libs/src/ownership/ownable.sw:45]: Good point. Instead of reading from memory again and creating a new variable, I now call `msg_sender()` since we've already verified that the sender is the owner.

File: sway_libs/src/ownership/ownable.sw
Comment (bitzoic) [reply sway_libs/src/ownership/ownable.sw:24]: This is definitely something I thought of when writing this. I would like to keep `renounce_ownership` as it follows openzeppilin's implementation of ownership and is something that Solidity developers are familiar with. 

Since this function should be called in the developer's constructor, in practice it shouldn't ever be called again... But it does leave the door open for them to shoot themselves in the foot.

I would also like to avoid the use of the 0 address as calling `owner()` should return `None`. Adding an enum to ensure that the `set_ownership()` function cannot be called twice is another approach but will incur more gas costs.

Would like to hear input from others as well on this

File: sway_libs/src/ownership/ownable.sw
Comment (simonr0204) [inline sway_libs/src/ownership/ownable.sw:15]: I think the 2nd condition here is sufficient.

**Comment by simonr0204** [comment]
Does this work? I didn't think libraries could read/write contract storage.

File: sway_libs/src/ownership/ownable.sw
Comment (simonr0204) [reply sway_libs/src/ownership/ownable.sw:24]: Non-initialized ownership and burned ownership are two distinct states - anyone using this library should be forced to handle them explicitly and separately. 

Maybe ownership could be an `enum` instead of an `option`, containing 3 variants for uninitialized, set, and burned ownership. We could provide some helper functions to `unwrap` the various cases.

Comment (matt-user) [reply sway_libs/src/ownership/ownable.sw:24]: I agree with @simonr0204 that we should have 3 variants for uninitialized, set, and renounced(burned) ownership.  Looking at the openzeppelin docs their implementation of `renounce_ownership` makes it so no one can set ownership after the function is called.  We should also have this behavior.

Comment (bitzoic) [reply sway_libs/src/ownership/ownable.sw:26]: I've updated the library accordingly

File: sway_libs/src/ownership/ownable.sw
Comment (bitzoic) [reply sway_libs/src/ownership/ownable.sw:17]: Correct me if I'm wrong but I believe this is needed. If the owner is `None` (In the cases of initialized or renounced) then we cannot unwrap it

**Review by matt-user — DISMISSED** [review]
lgtm but you may want to wait for another review

**Comment by simonr0204** [comment]
I was wondering whether it would be cleaner to combine the notions of `State` and `Owner` into something like:

```
pub enum Owner {
     Uninitialized: (),
     Initialized: Identity,
     Revoked: (),
 }
```

Then instead of `unwrap` on the owner, you would have a match statement on this enum.

**Comment by bitzoic** [comment]
> I was wondering whether it would be cleaner to combine the notions of `State` and `Owner` into something like:
> 
> ```
> pub enum Owner {
>      Uninitialized: (),
>      Initialized: Identity,
>      Revoked: (),
>  }
> ```
> 
> Then instead of `unwrap` on the owner, you would have to match on this enum, and handle each of the possible owner states. IMO this is fine (and desirable) since any app using this should be prepared to deal with the three possible states.
> 
> If we really wanted to return this as an option (i.e, force it into one of two states : Some or None) we could implement something like :
> 
> ```
> pub fn get_owner_as_option() -> Option<Identity> {
>      match owner() {
>          Owner::Initialized(identity) => Option::Some(identity),
>          _ => Option::None(),
>      }   
> )
> ```

This is wayyyy cleaner. I've updated it to use the identity in the enum

**Review by simonr0204 — APPROVED** [review]

**Review by matt-user — APPROVED** [review]

**Review by simonr0204 — APPROVED** [review]

**Pull request merged**
